<!--
  Waiterbot local navigation just with IR and bumber sensors:
  - turtlebot
  - move_base
  - fake localization
 -->

<launch>

  <!--  *********** Turtlebot and 3d sensor ***********  -->
  <include file="$(find turtlebot_bringup)/launch/minimal.launch"/>

  <!--  ************* Kobuki IR half ring *************  -->
  <include file="$(find waiterbot)/launch/includes/_ir_frames.launch"/>
  <node name="arduino_gateway" pkg="rosserial_python" type="serial_node.py" args="/dev/ttyACM0"/>
  <node name="ir_to_laserscan" pkg="waiterbot" type="ir_scan_node">
    <param name="closest_rejected" value="0.35"/>
    <param name="farthest_contact" value="0.8"/>
    <param name="readings_buffer" value="15"/>
    <param name="max_mean_error" value="10.0"/>
    <param name="max_covariance" value="0.005"/>
    <param name="noisy_readings" value="2"/>
  </node>

  <!--  ********** Kobuki mobile base extras **********  -->
  <include file="$(find turtlebot_bringup)/launch/includes/kobuki/_safety_controller.launch"/>
  <param name="cmd_vel_mux/yaml_cfg_file" value="$(find waiterbot)/param/vel_multiplexer.yaml"/>

  <!--  ******** Kobuki auto-docking controller *******  -->
  <include file="$(find kobuki_auto_docking)/launch/minimal.launch"/>

  <!--  ********** Alvar AR markers tracker ***********  -->
  <include file="$(find waiterbot)/launch/includes/_ar_tracker.launch">
    <arg name="max_frequency" value="4.0"/>  <!-- This is a CPU-hungry bastard; reduce its rate --> 
  </include>

  <!--  ************* Navigation manager **************  -->  
  <node name="waiterbot" pkg="waiterbot" type="waiterbot_node">
    <param name="base_beacon_distance" value="0.4"/>   <!-- At which distance from the base we start relaying on ir beacons -->  
    <param name="base_marker_distance" value="1.4"/>   <!-- At which distance from the base we start relaying on ar markers -->  
    <param name="check_localized" value="true"/>
    <param name="amcl_max_error" value="1.0"/>
    <param name="global_frame" value="map"/>
    <param name="odom_frame" value="odom"/>
    <param name="base_frame" value="base_footprint"/>
    <param name="tf_broadcast_freq" value="0.0"/>

    <!-- Incoming remaps -->
    <remap from="new_goal"    to="move_base_simple/goal"/>
    <remap from="amcl_init"   to="initialpose"/>
    <remap from="amcl_pose"   to="amcl_pose"/>
    <remap from="odometry"    to="odom"/>

    <!-- Outgoing remaps -->
    <remap from="issue_goal"  to="move_base_simple/goal"/>
    <remap from="cancel_goal" to="move_base/cancel"/>
    <remap from="reset_pose"  to="initialpose"/>

    <remap from="mobile_base/commands/velocity" to="cmd_vel_mux/input/special_tasks"/>
  </node>

  <!--  ************** Rosnav move base ***************  -->
  <include file="$(find waiterbot)/launch/includes/_move_base.launch"/>
  
  <!--
    Tweaks to adapt the move base node to local navigation:
     - global costmap is also a rolling window, non-static map
     - so it's referenced to odom, as the local costmap
     - and it must be updated, dislike a static map
     - we use a bigger costmap than that of the local planner
     - replan each 2 seconds, as global costmap changes
  -->
  <param name="move_base/global_costmap/global_frame" value="odom"/>
  <param name="move_base/global_costmap/update_frequency" value="1.0"/>
  <param name="move_base/global_costmap/publish_frequency" value="1.0"/>
  <param name="move_base/global_costmap/static_map" value="false"/>
  <param name="move_base/global_costmap/rolling_window" value="true"/>
  <param name="move_base/global_costmap/width" value="10.0"/>
  <param name="move_base/global_costmap/height" value="10.0"/>
  <param name="move_base/global_costmap/resolution" value="0.1"/>
  <param name="move_base/global_costmap/origin_x" value="0.0"/>
  <param name="move_base/global_costmap/origin_y" value="0.0"/>
  <param name="move_base/global_costmap/obstacle_range" value="5.0"/>
  <param name="move_base/global_costmap/raytrace_range" value="5.5"/>
  <param name="move_base/global_costmap/observation_sources" value="ir_scan bumpers"/>
  <param name="move_base/local_costmap/observation_sources" value="ir_scan bumpers"/>
  <param name="move_base/planner_frequency" value="0.5"/>

  <!--  ************** Fake Localization *************  -->
  <node name="fake_localization" pkg="fake_localization" type="fake_localization">
    <param name="odom_frame_id" value="/odom"/>
    <param name="global_frame_id" value="/map"/>
    <param name="base_frame_id" value="/base_footprint"/>
  </node>
</launch>
